### PR TITLE
Ph psgo 902 pull command

### PR DESCRIPTION
**Pull request by hosekpeter**
Jira: XXXXX

**Changes:**
- ...

-----------

### CODE REVIEW

File: internal/pkg/state/registry/registry.go [hunk 1]
```diff
@@ -14,6 +16,8 @@ import (
 
 type pathsRO = knownpaths.PathsReadOnly
 
+const KBCIgnoreFilePath = ".keboola/kbc_ignore"
```
Comment (Matovidlo) [inline internal/pkg/state/registry/registry.go:19]: Can we have separate module for ingnoring? Same as manifest or jsonnet files

File: internal/pkg/state/registry/registry.go [hunk 2]
```diff
@@ -261,3 +328,47 @@ func (s *Registry) GetOrCreateFrom(manifest ObjectManifest) (ObjectState, error)
 
 	return s.CreateFrom(manifest)
 }
+
+// applyIgnorePattern applies a single ignore pattern, marking the appropriate config or row as ignored.
+func (s *Registry) applyIgnorePattern(pattern string) error {
+	parts := strings.Split(pattern, "/")
+
+	switch len(parts) {
+	case 2:
```
Comment (Matovidlo) [inline internal/pkg/state/registry/registry.go:337]: this look magical, can we ignore basically any pattern that has any number of parts? 
For example we add new layer/ folder and we need to change the IgnorePattern.

File: internal/pkg/state/registry/registry_test.go [hunk 3]
```diff
@@ -301,3 +318,126 @@ func newTestState(t *testing.T, paths *knownpaths.Paths) *Registry {
...
+		},
+		{
+			name: "wrong pattern",
+			args: args{
+				pattern: "wrong pattern",
+			},
+			wantErr: assert.Error,
+		},
+		{
+			name: "too long pattern",
```
Comment (Matovidlo) [inline internal/pkg/state/registry/registry_test.go:416]: This looks like ok pattern, just changed dir structure. Could be changed in future

Comment (hosekpeter) [reply internal/pkg/state/registry/registry_test.go:416]: I will update documentation:

File: internal/pkg/project/ignore/file.go [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	f := &File{
+		state:            state,
+		rawStringPattern: content.Content,
+	}
+
+	return f, nil
```
Comment (Matovidlo) [inline internal/pkg/project/ignore/file.go:40]:
```suggestion
	return newFile(content.Content, state), nil
```

File: internal/pkg/project/ignore/ignore.go [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+// applyIgnoredPatterns parses the content for ignore patterns and applies them to configurations or rows.
+func (f *File) applyIgnoredPatterns() error {
+	for _, pattern := range f.parseIgnoredPatterns() {
+		if err := f.applyIgnorePattern(pattern); err != nil {
+			continue
+		}
+	}
+	return nil
+}
```
Comment (Matovidlo) [inline internal/pkg/project/ignore/ignore.go:55]:
```suggestion
// applyIgnoredPatterns parses the content for ignore patterns and applies them to configurations or rows.
func (f *File) applyIgnoredPatterns() error {
	for _, pattern := range f.parseIgnoredPatterns() {
		if err := f.applyIgnorePattern(pattern); err != nil {
			continue
		}
	}
	return nil
}

func (f *File) parseIgnoredPatterns() []string {
	var ignorePatterns []string
	lines := strings.Split(f.rawStringPattern, "\n")
	for _, line := range lines {
		trimmedLine := strings.TrimSpace(line)
		// Skip empty lines and comments
		if trimmedLine != "" && !strings.HasPrefix(trimmedLine, "#") {
			ignorePatterns = append(ignorePatterns, trimmedLine)
		}
	}

	return ignorePatterns
}

// applyIgnorePattern applies a single ignore pattern, marking the appropriate config or row as ignored.
func (f *File) applyIgnorePattern(ignoreConfig string) error {
	parts := strings.Split(ignoreConfig, "/")

	switch len(parts) {
	case 2:
		// Ignore config by ID and name.
		configID, componentID := parts[1], parts[0]
		f.state.IgnoreConfig(configID, componentID)
	case 3:
		// Ignore specific config row.
		configID, rowID := parts[1], parts[2]
		f.state.IgnoreConfigRow(configID, rowID)
	default:
		return errors.Errorf("invalid ignore ignoreConfig format: %s", ignoreConfig)
	}

	return nil
}
```
TL;DR the applyIgnoredPatterns should be on top

File: internal/pkg/state/registry/registry.go [hunk 6]
```diff
@@ -14,6 +14,8 @@ import (
 
 type pathsRO = knownpaths.PathsReadOnly
 
+const KBCIgnoreFilePath = ".keboola/kbc_ignore"
```
Comment (Matovidlo) [inline internal/pkg/state/registry/registry.go:17]:
```suggestion
```

File: internal/pkg/state/registry/registry.go [hunk 7]
```diff
@@ -183,6 +235,20 @@ func (s *Registry) ConfigRowsFrom(config ConfigKey) (rows []*ConfigRowState) {
...
+//	if err != nil {
+//		return err
+//	}
+//
+//	if content.Content == "" {
+//		return nil
+//	}
+//
+//	return s.applyIgnoredPatterns(content.Content)
+//}
```
Comment (Matovidlo) [inline internal/pkg/state/registry/registry.go:250]:
```suggestion
```

File: internal/pkg/state/registry/registry_test.go [hunk 8]
```diff
@@ -175,6 +175,21 @@ func TestStateTrackRecordInvalid(t *testing.T) {
...
+//	fs := aferofs.NewMemoryFs()
+//	require.NoError(t, fs.WriteFile(ctx, filesystem.NewRawFile(`foo/bar1`, "keboola.bar/678/34\nkeboola.foo/345")))
+//
+//	require.NoError(t, registry.SetIgnoredConfigsOrRows(ctx, fs, "foo/bar1"))
+//
+//	assert.Len(t, registry.IgnoredConfigRows(), 1)
+//	assert.Len(t, registry.IgnoredConfigs(), 1)
+//	assert.Equal(t, registry.IgnoredConfigRows()[0].ID.String(), "34")
+//	assert.Equal(t, registry.IgnoredConfigs()[0].ID.String(), "345")
+//}
```
Comment (Matovidlo) [inline internal/pkg/state/registry/registry_test.go:191]:
```suggestion
```

File: internal/pkg/state/registry/registry.go (hunk 7 above)
Comment (hosekpeter) [reply internal/pkg/state/registry/registry.go]: [2c14c77](https://github.com/keboola/keboola-as-code/pull/2136/commits/2c14c7783abac54d60b6f12e32664d6b2704cd74)

File: internal/pkg/state/registry/registry.go (hunk 6 above)
Comment (hosekpeter) [reply internal/pkg/state/registry/registry.go]: [2c14c77](https://github.com/keboola/keboola-as-code/pull/2136/commits/2c14c7783abac54d60b6f12e32664d6b2704cd74)

File: internal/pkg/project/ignore/ignore.go (hunk 5 above)
Comment (hosekpeter) [reply internal/pkg/project/ignore/ignore.go]: [2c14c77](https://github.com/keboola/keboola-as-code/pull/2136/commits/2c14c7783abac54d60b6f12e32664d6b2704cd74)

File: internal/pkg/project/ignore/file.go (hunk 4 above)
Comment (hosekpeter) [reply internal/pkg/project/ignore/file.go]: [2c14c77](https://github.com/keboola/keboola-as-code/pull/2136/commits/2c14c7783abac54d60b6f12e32664d6b2704cd74)

File: internal/pkg/state/registry/registry.go [hunk 9]
```diff
@@ -150,11 +154,19 @@ func (s *Registry) Configs() (configs []*ConfigState) {
...
+func (s *Registry) IgnoreConfig(ignoreID string, componentID string) {
 	for _, object := range s.All() {
 		if v, ok := object.(*ConfigState); ok {
-			if v.ID.String() == ignoreID {
+			if v.ID.String() == ignoreID && v.ComponentID.String() == componentID {
+				// ignore configuration
 				v.Ignore = true
+
+				// ignore rows of the configuration
+				if len(s.ConfigRowsFrom(v.ConfigKey)) > 0 {
```
Comment (jachym-tousek-keboola) [inline internal/pkg/state/registry/registry.go:161]: This condition seems useless?

File: pkg/lib/operation/project/sync/pull/operation.go [hunk 10]
```diff
@@ -118,3 +126,17 @@ func Run(ctx context.Context, projectState *project.State, o Options, d dependen
 
 	return nil
 }
+
+func ignoreConfigsAndRows(projectState *project.State) {
+	if len(projectState.IgnoredConfigRows()) > 0 {
```
Comment (jachym-tousek-keboola) [inline pkg/lib/operation/project/sync/pull/operation.go:138]: simlarly this one

File: pkg/lib/operation/project/sync/pull/operation.go [hunk 11]
```diff
@@ -118,3 +126,17 @@ func Run(ctx context.Context, projectState *project.State, o Options, d dependen
...
+		for _, v := range projectState.IgnoredConfigRows() {
+			v.SetRemoteState(nil)
+		}
+	}
+
+	if len(projectState.IgnoredConfigs()) > 0 {
```
Comment (jachym-tousek-keboola) [inline pkg/lib/operation/project/sync/pull/operation.go:144]: and this one

File: internal/pkg/state/registry/registry.go [hunk 12]
```diff
@@ -235,20 +233,6 @@ func (s *Registry) ConfigRowsFrom(config ConfigKey) (rows []*ConfigRowState) {
 	return rows
 }
 
-// func (s *Registry) SetIgnoredConfigsOrRows(ctx context.Context, fs filesystem.Fs, path string) error {
```
Comment (jachym-tousek-keboola) [inline internal/pkg/state/registry/registry.go]: Why is this commented out?

**Review by jachym-tousek-keboola — CHANGES_REQUESTED** [review]

File: internal/pkg/state/registry/registry.go (hunk 12 above)
Comment (hosekpeter) [reply internal/pkg/state/registry/registry.go]: [2c14c77](https://github.com/keboola/keboola-as-code/pull/2136/commits/2c14c7783abac54d60b6f12e32664d6b2704cd74)

Comment (hosekpeter) [reply internal/pkg/state/registry/registry.go]: it's removed

File: internal/pkg/state/registry/registry.go (hunk 9 above)
Comment (hosekpeter) [reply internal/pkg/state/registry/registry.go:161]: The configuration can have rows configuration. If you want to remove this configuration you need to remove rows too.

Comment (hosekpeter) [reply internal/pkg/state/registry/registry.go:161]: but you are right, this condition does not have to be there

File: pkg/lib/operation/project/sync/pull/operation.go (hunk 11 above)
Comment (hosekpeter) [reply pkg/lib/operation/project/sync/pull/operation.go]: [ae0cdbe](https://github.com/keboola/keboola-as-code/pull/2136/commits/ae0cdbebc3d40eb6f5c644e1c3c260a1c5d56030)

File: pkg/lib/operation/project/sync/pull/operation.go (hunk 10 above)
Comment (hosekpeter) [reply pkg/lib/operation/project/sync/pull/operation.go]: [ae0cdbe](https://github.com/keboola/keboola-as-code/pull/2136/commits/ae0cdbebc3d40eb6f5c644e1c3c260a1c5d56030)

File: internal/pkg/state/registry/registry.go (hunk 9 above)
Comment (hosekpeter) [reply internal/pkg/state/registry/registry.go]: [ae0cdbe](https://github.com/keboola/keboola-as-code/pull/2136/commits/ae0cdbebc3d40eb6f5c644e1c3c260a1c5d56030)

**Review by jachym-tousek-keboola — APPROVED** [review]

**Review by Matovidlo — APPROVED** [review]
